EyeSeeTea/dhis2-light-sdk
# Feature convert library to multiplatform

### :pushpin: References
* **Issue:** close https://github.com/EyeSeeTea/malariapp/issues/2203
* **Related pull-requests:** https://github.com/EyeSeeTea/malariapp/pull/2207

### :tophat: What is the goal?

Convert library to multiplatform with JVM and javascript as targets

###   :gear: branches 
**dhis2-light-sdk**:
       Origin: feature-feature-convert-library-to-multiplatform Target: developent

### :memo: How is it being implemented?

**Important!!**: Multiplatform is an experimental feature in kotlin

* First I have modified project structure to can open in Android Studio as an independent project
* I have used ktor to realize request to dhis2 API
* I have used ktor-engine to realize test with mocked responses
*  I have used kotlinx serialization to serialize server responses
*  I have created task to run test in js with mocha
* I have create task to create bundle.js in web sample using browserify
* I have created a new Datetime type because does not exists in common module any date type, this type is usually a platform dependent type (java.util.Date , js Date). The parse and format methods are created on every platform and its contract is created in the common module using expect keyword

**Important!!**: 
Generic serialization in javascript is broken for Js
For the moment we are going to create a D2CollectionResponse for every endpoint as OptionSetCollection, when this bug is solved we should use D2CollectionResponse to avoid create similar classes.

https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/custom_serializers.md#about-generic-serializers
https://github.com/Kotlin/kotlinx.serialization/issues/244

### :boom: How can it be tested?

clone dhis2-light-sdk and open as a stand-alone Android Studio project

 **Use case 1:** execute code analysis and execute all tests with this command:
```
./gradlew build
```
Ktlint is configured as a dependency of check subtask of build task

**Use case 2:** execute Android sample app, insert url, username, password and click on the button to realize request against dhis2 server. The response toString method should be rendered on the screen

**Use case 3:** execute Web sample

1. Open chrome disabling web security:

For OSX:
```
$ open -a Google\ Chrome --args --disable-web-security --user-data-dir
```
--user-data-dir required on Chrome 49+ on OSX

For Linux run:
```
$ google-chrome --disable-web-security
```

For Windows go into the command prompt and go into the folder where Chrome.exe is and type:
```
chrome.exe --disable-web-security
```

2. Compile code to js and create bundle.js

execute:
```
./gradlew buildJsToWeb
```
3. Open web sample
Open index.html under webSample directory, insert URL, username, password and click on the button to realize request against dhis2 server. The response toString method should be rendered on the screen

### :floppy_disk: Requires DB migration?


### :art: UI changes?


<img width="200" alt="android-sample-app" src="https://user-images.githubusercontent.com/5593590/49922975-737ce980-feb2-11e8-9967-2da35537413c.png">
<img width="400" alt="web-sample-app" src="https://user-images.githubusercontent.com/5593590/49922982-78da3400-feb2-11e8-8bd4-fc621e588eca.png">
